### PR TITLE
[inductor] Parallelize Max Autotune step 1: Use Popen

### PR DESCRIPTION
**Pull request by masnesral**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #107983
* __->__ #107982

Summary: Step 1 in revamping subprocess autotune to support multiple GPUs: use Popen to create a new process with an entry point we control so we don't reinterpret the toplevel script.

Test Plan: `python test/inductor/test_max_autotune.py`

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

### CODE REVIEW

File: torch/_inductor/autotune_process.py
Comment (Skylion007) [inline torch/_inductor/autotune_process.py:266]: Please specify stack level of warnings.warn.

File: torch/_inductor/autotune_process.py
Comment (Skylion007) [inline torch/_inductor/autotune_process.py:45]: Just add this as a logger call instead of a temporary print statement.

Comment (masnesral) [reply torch/_inductor/autotune_process.py:45]: @Skylion007, yeah these print stmts were already here and I just changed them to dump to stderr since this refactor uses stdout for the IPC. So printing to stdout would mess up the messaging. Arguably these prints were never going to be helpful in the first place, however, because the sub-process doesn't have a terminal to print them. I can just remove them since I'm not sure any of the 3 instances are terribly useful ... Or, the "right" approach is probably to dump to a file. If you recommend that approach, do you know if this code base has a canonical pattern for writing debug info to a tmp file?

File: torch/_inductor/autotune_process.py
Comment (Skylion007) [inline torch/_inductor/autotune_process.py:117]:
```suggestion
        except pickle.UnpicklingError as ex:
            raise RuntimeError(
                "Error deserializing response from the benchmarking subprocess. "
                "Is the benchmark code path writing to stdout?"
            ) from ex
```

File: torch/_inductor/autotune_process.py
Comment (Skylion007) [reply torch/_inductor/autotune_process.py:266]: Yeah, we need to fix this in the codebase to improve the warning levels: https://docs.python.org/3/library/warnings.html#warnings.warn the stack_level kwarg should be greater than the default because the default is basically useless (should be at least 2 so people can figure out where it's called from)

File: torch/_inductor/autotune_process.py
Comment (eellison) [inline torch/_inductor/autotune_process.py:112]: for my own understanding, is there a reason why we initialize before re-throwing ?

File: torch/_inductor/autotune_process.py
Comment (eellison) [inline torch/_inductor/autotune_process.py:218]: it seems a bit off to log this to stderr. is this to avoid parsing it back on the main thread ?

**Review by eellison — COMMENTED** [review]
looks great! just had a couple comments. it might be worth putting in code comments somewhere about the basics of how this works

File: torch/_inductor/autotune_process.py
Comment (masnesral) [reply torch/_inductor/autotune_process.py:112]: @eellison: I'm sure we could do it elsewhere if it makes more sense. It made sense to me to keep the management of the sub-process wholly contained within this class (i.e., the TuningProcess object makes sure its sub-process is always running). This is the most immediate point where we know the sub-process exited, so I just restart it immediately. Another approach is to re-initialize, if necessary, in the put() method. Or get even more elaborate and figure out how to reinitialize asynchronously so we don't block on the sub-process creation?

File: torch/_inductor/autotune_process.py
Comment (masnesral) [reply torch/_inductor/autotune_process.py:218]: @eellison, yeah, see my comment above to skylion007. These print stmts were already going nowhere and I just changed them to stderr in case anyone ever enabled DEBUG above (but they'll still go nowhere). Accidentally writing to stdout would mess up the message passing between processes and is pretty confusing. If we want to clean up the logging in this diff, I'd suggest either 1) having the sub-process log to a tmp file somewhere, or 2) We can continue to log to stderr (but probably using a real logger), read the stderr from the pipe in the main process, and log it on the sub-process's behalf. What do you think?

File: torch/_inductor/autotune_process.py
Comment (masnesral) [reply torch/_inductor/autotune_process.py:45]: ellison has a similar comment; I'll adding logging to a file

File: torch/_inductor/autotune_process.py
Comment (masnesral) [reply torch/_inductor/autotune_process.py:218]: I'll go the route of logging to a file

File: torch/_inductor/autotune_process.py
Comment (shunting314) [reply torch/_inductor/autotune_process.py:45]: My bad using a DEBUG flag in the first place. Back then we don't have TORCH_LOGS support ..

Does using logging.Logger.debug work here? Those message should by default go to stderr

File: torch/_inductor/autotune_process.py
Comment (shunting314) [inline torch/_inductor/autotune_process.py:73]: Let me take a step back. Would this be a bit fragile. Any extra print to stdout in child process would mess things up?

File: torch/_inductor/autotune_process.py
Comment (shunting314) [inline torch/_inductor/autotune_process.py:47]: Any reason we can not reuse put method below?

```
   def put(self, obj: Any) -> None:
        """
        Push a work item to the child process.
        """
        assert self.process is not None
        assert self.process.stdin is not None
        pickle.dump(obj, self.process.stdin)
        self.process.stdin.flush()
```

File: torch/_inductor/autotune_process.py
Comment (eellison) [reply torch/_inductor/autotune_process.py:73]: Within the subprocess, we could capture stdout when we are not explicitly communicating

File: torch/_inductor/autotune_process.py
Comment (shunting314) [reply torch/_inductor/autotune_process.py:47]: NVM, they are different. Might be good to change this as a method together with put/get methods below since they are very similar

File: torch/_inductor/autotune_process.py
Comment (shunting314) [reply torch/_inductor/autotune_process.py:112]: Just FYI, remove `self.initialize` here is fine since we call `self.initialize` in benchmark_in_sub_proc` anyways. Removing the call here makes the sub-process recreation lazy.

File: torch/_inductor/autotune_process.py
Comment (masnesral) [reply torch/_inductor/autotune_process.py:45]: @shunting314, no problem. I'm happy to change it. Just wasn't sure how much to evolve the implementation in one PR. The problem with logging normally to stderr is that this code runs in the sub-process, and there's no console. I was gonna take a crack at just logging to a file instead. I think that might be the most usable especially after the next diff in the stack where we have multiple independent sub-processes. What do you think?

**Comment by shunting314** [comment]
Any noticeable perf difference by using Popen?

File: torch/_inductor/autotune_process.py
Comment (masnesral) [reply torch/_inductor/autotune_process.py:73]: Yes, it would be vulnerable if anyone adds a new print stmt. I was gonna change to use a real logger instead and have that log to a file?

File: torch/_inductor/autotune_process.py
Comment (shunting314) [reply torch/_inductor/autotune_process.py:45]: hmm, didn't subprocess inherits stderr from parent process? Logging to a separate file at lease makes it hard to check output in senarios like CI. And people need to be aware there is a separate file to check. If stderr works, I'll prefer stderr.


And yea, this can be done in separate PR

Comment (masnesral) [reply torch/_inductor/autotune_process.py:45]: > hmm, didn't subprocess inherits stderr from parent process?

I couldn't get it to work out of the box, but as Elias might tell you, I'm kind of a noob working with python sub-processes ;)  I'll look again and figure it out.

File: torch/_inductor/autotune_process.py
Comment (masnesral) [reply torch/_inductor/autotune_process.py:112]: Sure, we can do lazy

**Comment by masnesral** [comment]
> Any noticeable perf difference by using Popen?

@shunting314, no I don't discern a difference in the benchmarking results at least. Or do you mean aggregate end-to-end execution time of the process?

**Comment by eellison** [comment]
@shunting314 do you have any repros for the slow down when auto tune in sub process was enabled ?

**Comment by masnesral** [comment]
@shunting314, FYI here's a side by side comparison using the hf_Bart benchmark

```
							      >	+ TORCHINDUCTOR_AUTOTUNE_IN_SUBPROC=1
+ TORCHINDUCTOR_MAX_AUTOTUNE=1					+ TORCHINDUCTOR_MAX_AUTOTUNE=1
+ python benchmarks/dynamo/torchbench.py --device cuda --perf	+ python benchmarks/dynamo/torchbench.py --device cuda --perf
loading model: 0it [00:09, ?it/s]				loading model: 0it [00:09, ?it/s]
cuda eval  hf_Bart						cuda eval  hf_Bart
AUTOTUNE mm(512x768, 768x768)					AUTOTUNE mm(512x768, 768x768)
  mm 0.0148 ms 100.0%					      |	  mm 0.0141 ms 100.0%
  triton_mm_8 0.0154 ms 96.5%				      |	  triton_mm_8 0.0141 ms 99.5%
  triton_mm_3 0.0164 ms 90.8%				      |	  triton_mm_3 0.0161 ms 87.3%
  triton_mm_4 0.0171 ms 86.9%				      |	  triton_mm_4 0.0163 ms 86.4%
  triton_mm_9 0.0173 ms 85.6%				      |	  triton_mm_9 0.0167 ms 84.5%
  triton_mm_6 0.0175 ms 84.7%				      |	  triton_mm_6 0.0175 ms 80.3%
  triton_mm_5 0.0181 ms 82.2%				      |	  triton_mm_5 0.0179 ms 78.5%
  triton_mm_1 0.0197 ms 75.4%				      |	  triton_mm_1 0.0203 ms 69.5%
  triton_mm_2 0.0204 ms 72.7%				      |	  triton_mm_2 0.0203 ms 69.3%
  triton_mm_0 0.0300 ms 49.6%				      |	  triton_mm_0 0.0301 ms 46.8%
SingleProcess AUTOTUNE takes 5.1835 seconds		      |	SubProcess AUTOTUNE takes 4.7125 seconds
AUTOTUNE bmm(12x512x64, 12x64x512)				AUTOTUNE bmm(12x512x64, 12x64x512)
  triton_bmm_26 0.0138 ms 100.0%			      |	  triton_bmm_31 0.0132 ms 100.0%
  triton_bmm_25 0.0138 ms 99.5%				      |	  triton_bmm_26 0.0133 ms 99.8%
  triton_bmm_28 0.0146 ms 94.1%				      |	  triton_bmm_25 0.0136 ms 97.2%
  triton_bmm_31 0.0146 ms 94.1%				      |	  triton_bmm_32 0.0137 ms 96.5%
  triton_bmm_27 0.0147 ms 93.5%				      |	  triton_bmm_28 0.0139 ms 95.2%
  triton_bmm_32 0.0147 ms 93.5%				      |	  bmm 0.0148 ms 89.8%
  bmm 0.0148 ms 93.2%					      |	  triton_bmm_27 0.0154 ms 86.1%
  triton_bmm_24 0.0151 ms 90.9%				      |	  triton_bmm_24 0.0156 ms 84.7%
  triton_bmm_34 0.0164 ms 84.0%				      |	  triton_bmm_34 0.0169 ms 78.6%
  triton_bmm_33 0.0182 ms 75.4%				      |	  triton_bmm_33 0.0173 ms 76.5%
SingleProcess AUTOTUNE takes 4.6074 seconds		      |	SubProcess AUTOTUNE takes 1.0350 seconds
AUTOTUNE bmm(12x512x512, 12x512x64)				AUTOTUNE bmm(12x512x512, 12x512x64)
  triton_bmm_51 0.0154 ms 100.0%			      |	  triton_bmm_56 0.0141 ms 100.0%
  triton_bmm_56 0.0154 ms 99.6%				      |	  triton_bmm_51 0.0147 ms 96.3%
  triton_bmm_57 0.0155 ms 99.0%				      |	  triton_bmm_57 0.0154 ms 91.7%
  triton_bmm_54 0.0161 ms 95.4%				      |	  triton_bmm_53 0.0157 ms 90.0%
  triton_bmm_53 0.0162 ms 95.0%				      |	  triton_bmm_49 0.0160 ms 88.0%
  triton_bmm_52 0.0163 ms 94.5%				      |	  triton_bmm_52 0.0162 ms 87.0%
  bmm 0.0166 ms 92.3%					      |	  triton_bmm_54 0.0163 ms 86.6%
  triton_bmm_49 0.0174 ms 88.4%				      |	  bmm 0.0166 ms 85.0%
  triton_bmm_50 0.0175 ms 87.8%				      |	  triton_bmm_50 0.0179 ms 78.7%
  triton_bmm_48 0.0244 ms 62.8%				      |	  triton_bmm_48 0.0238 ms 59.3%
SingleProcess AUTOTUNE takes 4.2848 seconds		      |	SubProcess AUTOTUNE takes 0.7750 seconds
AUTOTUNE mm(512x768, 768x3072)					AUTOTUNE mm(512x768, 768x3072)
  mm 0.0214 ms 100.0%					      |	  mm 0.0206 ms 100.0%
  triton_mm_74 0.0230 ms 92.8%				      |	  triton_mm_74 0.0230 ms 89.9%
  triton_mm_73 0.0231 ms 92.6%				      |	  triton_mm_73 0.0238 ms 86.6%
  triton_mm_76 0.0251 ms 85.1%				      |	  triton_mm_76 0.0244 ms 84.6%
  triton_mm_75 0.0252 ms 85.0%				      |	  triton_mm_75 0.0246 ms 83.8%
  triton_mm_80 0.0296 ms 72.2%				      |	  triton_mm_80 0.0292 ms 70.6%
  triton_mm_72 0.0338 ms 63.3%				      |	  triton_mm_72 0.0328 ms 62.9%
  triton_mm_79 0.0377 ms 56.7%				      |	  triton_mm_79 0.0383 ms 53.9%
  triton_mm_82 0.0420 ms 50.9%				      |	  triton_mm_82 0.0416 ms 49.6%
  triton_mm_81 0.0465 ms 46.0%				      |	  triton_mm_81 0.0456 ms 45.3%
SingleProcess AUTOTUNE takes 4.7278 seconds		      |	SubProcess AUTOTUNE takes 0.8443 seconds
AUTOTUNE mm(512x3072, 3072x768)					AUTOTUNE mm(512x3072, 3072x768)
  mm 0.0277 ms 100.0%					      |	  mm 0.0283 ms 100.0%
  triton_mm_92 0.0358 ms 77.4%				      |	  triton_mm_92 0.0363 ms 78.0%
  triton_mm_87 0.0420 ms 66.1%				      |	  triton_mm_88 0.0414 ms 68.4%
  triton_mm_88 0.0420 ms 66.1%				      |	  triton_mm_87 0.0419 ms 67.6%
  triton_mm_89 0.0440 ms 63.0%				      |	  triton_mm_89 0.0440 ms 64.3%
  triton_mm_90 0.0451 ms 61.6%				      |	  triton_mm_90 0.0445 ms 63.6%
  triton_mm_93 0.0461 ms 60.2%				      |	  triton_mm_93 0.0462 ms 61.3%
  triton_mm_86 0.0563 ms 49.3%				      |	  triton_mm_86 0.0566 ms 50.1%
  triton_mm_85 0.0567 ms 49.0%				      |	  triton_mm_85 0.0566 ms 50.0%
  triton_mm_84 0.0708 ms 39.2%				      |	  triton_mm_84 0.0693 ms 40.9%
SingleProcess AUTOTUNE takes 4.5848 seconds		      |	SubProcess AUTOTUNE takes 1.1149 seconds
AUTOTUNE mm(512x768, 768x50265)					AUTOTUNE mm(512x768, 768x50265)
  triton_mm_1585 0.2616 ms 100.0%			      |	  triton_mm_1585 0.2570 ms 100.0%
  triton_mm_1586 0.2622 ms 99.8%			      |	  triton_mm_1586 0.2654 ms 96.8%
  triton_mm_1591 0.3036 ms 86.2%			      |	  triton_mm_1591 0.3005 ms 85.5%
  triton_mm_1587 0.3041 ms 86.0%			      |	  triton_mm_1587 0.3045 ms 84.4%
  triton_mm_1588 0.3084 ms 84.8%			      |	  triton_mm_1588 0.3071 ms 83.7%
  triton_mm_1584 0.3270 ms 80.0%			      |	  triton_mm_1584 0.3268 ms 78.6%
  triton_mm_1592 0.4618 ms 56.6%			      |	  triton_mm_1592 0.4638 ms 55.4%
  triton_mm_1594 0.4884 ms 53.6%			      |	  triton_mm_1594 0.4878 ms 52.7%
  triton_mm_1593 0.5923 ms 44.2%			      |	  triton_mm_1593 0.5705 ms 45.0%
  triton_mm_1590 0.6220 ms 42.1%			      |	  triton_mm_1590 0.6091 ms 42.2%
SingleProcess AUTOTUNE takes 5.7595 seconds		      |	SubProcess AUTOTUNE takes 1.0035 seconds
running benchmark: 100%|█████████████████████████████████████ |	running benchmark: 100%|█████████████████████████████████████
2.971x							      |	2.950x
```

**Comment by shunting314** [comment]
> do you have any repros for the slow down when auto tune in sub process was enabled ?

I tried on cmf model before and sub-process tuning cause about 20-30% slowdown IIRC. While leveraging multi-GPU can brings some speedup (although not linear).

I think there are 2 kinds of perf tests we can do here:
1. test on some micro benchmark. E.g, a matmul kernel
2. test on existing benchmarks (TB/HF/TIMM) with max-autotune enabled.

File: torch/_inductor/autotune_process.py
Comment (shunting314) [inline torch/_inductor/autotune_process.py:29]: We don't need this. By default, log level is set to WARNING. If you want to enable debug log for a module like this, override TORCH_LOGS environment variable should work:
```
TORCH_LOGS=+torch._inductor.autotune_process python xxx
```

Comment (masnesral) [reply torch/_inductor/autotune_process.py:29]: Ah. Thanks!

**Review by eellison — APPROVED** [review]
Looks good to me. I think we should land these in a stack. And let's wait for @shunting314 to approve as well.

In a follow up pr, let's make this more robust to spurious prints ?

**Review by shunting314 — APPROVED** [review]
So previously we use explicit Queue to communicate between parent and child process. Now we use STDIN and STDOUT of the child process.  The former is like dedicated IPC channels while the latter can be accidentally shared by other code.

Can we continue to have dedicated IPC channels that either via Queue or via anything else similar. Accept to unblock.

**Pull request closed without merging**
